File: sample_app_dagger/src/main/java/com/chumarin/stanislav/sample_app_dagger/SessionTrackerListener.java
package com.chumarin.stanislav.sample_app_dagger;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.chumarin.stanislav.sample_app_dagger.di.UserComponent;
import com.chumarin.stanislav.sample_app_dagger.di.UserModule;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.jetbrains.annotations.NotNull;
import vit.khudenko.android.sessiontracker.SessionRecord;
import vit.khudenko.android.sessiontracker.SessionTracker;

public class SessionTrackerListener implements SessionTracker.Listener<Session.Event, Session.State> {

    private static final String TAG = SessionTrackerListener.class.getSimpleName();
    @NonNull
    private final App app;

    @NonNull
    private final Map<String, UserComponent> sessionUserComponents = new HashMap<>();

    public SessionTrackerListener(@NonNull App app) {
        this.app = app;
    }

    private void closeDaggerScope(@NonNull String sessionId) {
        if (sessionUserComponents.remove(sessionId) != null) {
            Log.d(TAG, "closeDaggerScope: session ID = '" + sessionId + "'");
        } else {
            Log.d(TAG, "closeDaggerScope: no scope to close for session with ID '" + sessionId + "'");
        }
    }

    private void createDaggerScope(@NonNull String sessionId) {
        if (sessionUserComponents.get(sessionId) != null) {
            Log.w(TAG, "createDaggerScope: scope already exists for session with ID '" + sessionId + "'");
        } else {
            Log.d(TAG, "createDaggerScope: session ID = '" + sessionId + "'");
            sessionUserComponents.put(sessionId, app.getAppComponent().attachComponent(new UserModule(sessionId)));
        }
    }

    @Override
    public void onSessionTrackerInitialized(@NotNull SessionTracker<Session.Event, Session.State> sessionTracker,
                                            @NotNull List<SessionRecord<Session.State>> sessionRecords) {
        Log.d(TAG, "onSessionTrackerInitialized");
        if (sessionRecords.stream().filter(sessionRecord -> sessionRecord.getState() == Session.State.ACTIVE).count() > 1) {
            throw new IllegalStateException("One active session is allowed at most");
        }
        for (SessionRecord<Session.State> sessionRecord : sessionRecords) {
            if (sessionRecord.getState() == Session.State.ACTIVE) {
                createDaggerScope(sessionRecord.sessionId());
            }
        }
    }

    @Override
    public void onSessionTrackingStarted(@NotNull SessionTracker<Session.Event, Session.State> sessionTracker,
                                         @NotNull SessionRecord<Session.State> sessionRecord) {
        String sessionId = sessionRecord.sessionId();
        Session.State state = sessionRecord.getState();
        Log.d(TAG, "onSessionTrackingStarted: session ID = " + sessionId + ", state = " + state);
        if (state == Session.State.ACTIVE) {
            createDaggerScope(sessionId);
        }
    }

    @Override
    public void onSessionStateChanged(@NotNull SessionTracker<Session.Event, Session.State> sessionTracker,
                                      @NotNull SessionRecord<Session.State> sessionRecord,
                                      @NotNull Session.State oldState) {
        String sessionId = sessionRecord.sessionId();
        Session.State newState = sessionRecord.getState();
        Log.d(TAG, "onSessionStateChanged: session ID = " + sessionId + ", states = (" + oldState + " -> " + newState + ")");
        switch (newState) {
            case ACTIVE:
                createDaggerScope(sessionId);
                break;
            case INACTIVE:
            case FORGOTTEN:
                closeDaggerScope(sessionId);
                break;
            default:
                throw new IllegalArgumentException("Unknown newState [" + newState + "]");
        }
    }

    @Override
    public void onSessionTrackingStopped(@NotNull SessionTracker<Session.Event, Session.State> sessionTracker,
                                         @NotNull SessionRecord<Session.State> sessionRecord) {
        String sessionId = sessionRecord.sessionId();
        Session.State state = sessionRecord.getState();
        Log.d(TAG, "onSessionTrackingStopped: session ID = " + sessionId + ", state = " + state);
        closeDaggerScope(sessionId);
    }

    @Nullable
    public UserComponent getUserComponent(String sessionId) {
        return sessionUserComponents.get(sessionId);
    }

    @Override
    public void onAllSessionsTrackingStopped(
            @NotNull SessionTracker<Session.Event, Session.State> sessionTracker,
            @NotNull List<SessionRecord<Session.State>> sessionRecords) {
        Log.d(TAG, "onAllSessionsTrackingStopped");
        for (SessionRecord<Session.State> sessionRecord : sessionRecords) {
            closeDaggerScope(sessionRecord.sessionId());
        }
    }
}

File: sample_app_koin/src/main/kotlin/vit/khudenko/android/sessiontracker/sample/koin/SessionTrackerListener.kt
package vit.khudenko.android.sessiontracker.sample.koin

import android.util.Log
import org.koin.core.Koin
import org.koin.core.qualifier.named
import org.koin.core.scope.Scope
import vit.khudenko.android.sessiontracker.SessionId
import vit.khudenko.android.sessiontracker.SessionRecord
import vit.khudenko.android.sessiontracker.SessionTracker
import vit.khudenko.android.sessiontracker.sample.koin.Session.Event
import vit.khudenko.android.sessiontracker.sample.koin.Session.State
import vit.khudenko.android.sessiontracker.sample.koin.di.SCOPE_ID_USER_SESSION

class SessionTrackerListener(private val koinProvider: () -> Koin) : SessionTracker.Listener<Event, State> {

    companion object {
        private val TAG = SessionTrackerListener::class.java.simpleName
    }

    private var scope: Scope? = null

    override fun onSessionTrackerInitialized(
        sessionTracker: SessionTracker<Event, State>,
        sessionRecords: List<SessionRecord<State>>
    ) {
        Log.d(TAG, "onSessionTrackerInitialized")
        check(sessionRecords.count { (_, state) -> state == State.ACTIVE } <= 1) {
            "One active session is allowed at most"
        }
        sessionRecords.forEach { (sessionId, state) ->
            if (state == State.ACTIVE) {
                createKoinScope(sessionId)
            }
        }
    }

    override fun onSessionTrackingStarted(
        sessionTracker: SessionTracker<Event, State>,
        sessionRecord: SessionRecord<State>
    ) {
        val (sessionId, initState) = sessionRecord
        Log.d(TAG, "onSessionTrackingStarted: session ID = ${sessionId.value}, initState = $initState")
        if (initState == State.ACTIVE) {
            createKoinScope(sessionId)
        }
    }

    override fun onSessionTrackingStopped(
        sessionTracker: SessionTracker<Event, State>,
        sessionRecord: SessionRecord<State>
    ) {
        val (sessionId, state) = sessionRecord
        Log.d(TAG, "onSessionTrackingStopped: session ID = ${sessionId.value}, state = $state")
        closeKoinScope(sessionId)
    }

    override fun onSessionStateChanged(
        sessionTracker: SessionTracker<Event, State>,
        sessionRecord: SessionRecord<State>,
        oldState: State
    ) {
        val (sessionId, newState) = sessionRecord
        Log.d(TAG, "onSessionStateChanged: session ID = ${sessionId.value}, states = ($oldState -> $newState)")
        when (newState) {
            State.ACTIVE -> createKoinScope(sessionId)
            State.INACTIVE,
            State.FORGOTTEN -> closeKoinScope(sessionId)
        }
    }

    override fun onAllSessionsTrackingStopped(
        sessionTracker: SessionTracker<Event, State>,
        sessionRecords: List<SessionRecord<State>>
    ) {
        Log.d(TAG, "onAllSessionsTrackingStopped")
        sessionRecords.forEach { (sessionId, _) -> closeKoinScope(sessionId) }
    }

    private fun closeKoinScope(sessionId: SessionId) {
        if (scope != null) {
            Log.d(TAG, "closeKoinScope: session ID = '${sessionId.value}'")
            scope!!.close()
            scope = null
        } else {
            Log.d(TAG, "closeKoinScope: no scope to close for session with ID '${sessionId.value}'")
        }
    }

    private fun createKoinScope(sessionId: SessionId) {
        if (scope != null) {
            Log.w(TAG, "createKoinScope: scope already exists for session with ID '${sessionId.value}'")
        } else {
            Log.d(TAG, "createKoinScope: session ID = '${sessionId.value}'")
            scope = koinProvider().createScope(
                qualifier = named(SCOPE_ID_USER_SESSION),
                scopeId = sessionId.value
            )
        }
    }
}